internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestEntries = 1000
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries map[string][]lokiEntry
+	// Keep track of the number of buffered entries, to avoid having to iterate
+	// the entries map to get the total number.
+	numEntries int
+
+	httpClient *http.Client
+	retryAfter *time.Time
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: &http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], encodeEntry(entry))
+	c.numEntries++
+
+	if c.numEntries >= maxRequestEntries {
+		if c.retryAfter != nil {
+			if time.Now().Before(*c.retryAfter) {
+				// Retry-After deadline hasn't passed yet, so we shouldn't flush
+				return nil
+			}
+			c.retryAfter = nil
+		}
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *Client) Flush(ctx context.Context) error {
+	if c.numEntries == 0 {
+		return nil // no-op
+	}
+	defer c.emptyBuffer()
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return handleHTTPClientErr(err)
+	}
+	return c.handleServerResponse(resp)
+}
+
+func (c *Client) emptyBuffer() {
+	for svc := range c.entries {
+		c.entries[svc] = c.entries[svc][:0]
+	}
+	c.numEntries = 0
+}
+
+func (c *Client) buildRequest() request {
+	// Sort keys to guarantee deterministic output
+	var services []string
+	for svc, entries := range c.entries {
+		if len(entries) == 0 {
+			continue
+		}
+		services = append(services, svc)
+	}
+	sort.Strings(services)
+
+	var req request
+	for _, service := range services {
+		entries := c.entries[service]
+		stream := stream{
+			Labels: map[string]string{
+				"pebble_service": service,
+			},
+			Entries: entries,
+		}
+		req.Streams = append(req.Streams, stream)
+	}
+	return req
+}
+
+type request struct {
+	Streams []stream `json:"streams"`
+}
+
+type stream struct {
+	Labels  map[string]string `json:"stream"`
+	Entries []lokiEntry       `json:"values"`
+}
+
+type lokiEntry [2]string
+
+func handleHTTPClientErr(err error) error {
+	var netErr net.Error
+	if errors.As(err, &netErr) && netErr.Timeout() {
+		return &clienterr.Timeout{Err: err}
```

Comment:
I think with Timeout being a const error it works better as you can enrich the error with more detail here.

Reply:
How would that work?

Reply:
Would allow us to write with `fmt.Errorf` and control the exact wording. At the moment because it's a struct error that is generating it's own error string we are at the mercy of the exact placement of words and stops us from being able to construct a better worded error for every case this might be used.

Small knit I just think a ConstError will work 10 times better here if it ends up staying around.

Reply:
This discussion is moot now because I removed `Timeout`.

However, my understanding is there are [two major paradigms](https://stackoverflow.com/a/76918940/15219560) for errors in Go:
1. const / sentinel errors - match with `errors.Is`
2. typed struct errors - match with `errors.As`

I don't want to mix and match these two different styles, as I think it's confusing for users of this package. And we already have to use (2) for some errors (e.g. `Backoff`) as we have to attach extra data to the error. Therefore, for consistency, I want to use (2) for all errors from this package - even though (1) might make sense in this particular case.

Reply:
Agreed the discussion is moot now that `Timeout` is gone. I think @barrettj12's explanation about the "two major paradigms" and your rationale about not wanting to mix the two different styles is very reasonable.